### ISSUE TITLE
What is the behavior if producer_partition_by=primary_key & table does not have primary key?

### ISSUE THREAD

**Issue by tankchintan** [zendesk/maxwell#555]
Do we know what is the behavior if in Kafka producer configuration we specify `producer_partition_by=primary_key` but the table does not have a primary key?

I know their is a `kafka_partiton_by_fallback` configuration, but that seems to be applicable only if we have `producer_partition_by=column`.

Also what would happen if the primary key is a composite key?

**Comment by osheroff** [comment]
https://github.com/zendesk/maxwell/blob/0ae24f107eeb16a8a69c2548d6a86dfd86c91f4f/src/main/java/com/zendesk/maxwell/row/RowMap.java#L137 shows that in the case of no primary key, maxwell falls back to partitioning by database+table.  Not sure if that's ideal; I think originally I had intended it to generate a GUID and basically partition-by-random.  Not really sure exactly what the right behavior is.

Composite keys are partitioned by a concatenation of the values of the primary key.

**Comment by tankchintan** [comment]
@osheroff Thanks for the quick turnaround. I am glad that there is some fallback. Although in my specific case, since I am planning to have one `Topic` per `table`, it might be moot for it to fallback on `db + table`. But I guess that's fine.

On a related note, I know that in Kafka when you try to publish to a non-existing `Topic` it can create one for you by specifying some config. But in that case do you know how many partitions will be created?

Or your recommendation is to always have `Topic` created before we start using maxwell?

**Comment by osheroff** [comment]
yeah I don't think num-partitions is configurable; I think it's always 1.  You could ask on the kafka mailing list, but right now highly recommended that you pre-create the topics.  When the topic-per-table PR was merged I poked at the idea of integrating some kafka-admin stuff into maxwell (a stripped down version of the toolchain that could auto-create topics with pre-configured settings), but it looked too big at the time.

**Comment by osheroff** [comment]
yup the docs are PR-able.
